### PR TITLE
Add `Future` and `Experimental` forks and networks

### PR DESCRIPTION
**Pull request by shemnon**
## PR description

Add two new development networks and forks, `Future` for EIPs not yet scheduled for a fork and `Experimental` for EIPs and other work not yet accepted to a public network.  Two new fork-named networks are added like the `dev` network to facilitate easy local development. These replace the deprecated `Shandong` network and fork.

Signed-off-by: Danno Ferrin <danno.ferrin@swirldslabs.com>

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: config/src/test/java/org/hyperledger/besu/config/GenesisConfigOptionsTest.java [hunk 1]
```diff
@@ -238,7 +244,11 @@ public void shouldNotReturnEmptyOptionalWhenBlockNumberNotSpecified() {
     assertThat(config.getLondonBlockNumber()).isEmpty();
     assertThat(config.getArrowGlacierBlockNumber()).isEmpty();
     assertThat(config.getGrayGlacierBlockNumber()).isEmpty();
-    assertThat(config.getShandongBlockNumber()).isEmpty();
+    assertThat(config.getMergeNetSplitBlockNumber()).isEmpty();
+    assertThat(config.getShanghaiTime()).isEmpty();
+    assertThat(config.getCancunTime()).isEmpty();
+    assertThat(config.getExperimentalTime()).isEmpty();
+    assertThat(config.getFutureTime()).isEmpty();
```
Comment (siladu) [inline config/src/test/java/org/hyperledger/besu/config/GenesisConfigOptionsTest.java:251]: nit: ordering - future then experimental

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecFactory.java [hunk 2]
```diff
@@ -207,8 +207,20 @@ public ProtocolSpecBuilder cancunDefinition(final GenesisConfigOptions genesisCo
         evmConfiguration);
   }
 
-  public ProtocolSpecBuilder shandongDefinition(final GenesisConfigOptions genesisConfigOptions) {
-    return MainnetProtocolSpecs.shandongDefinition(
+  public ProtocolSpecBuilder futureDefinition(final GenesisConfigOptions genesisConfigOptions) {
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecFactory.java:210]: Might be good to javadoc how to decide whether features should go under future or experimental?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolScheduleBuilder.java [hunk 3]
```diff
@@ -168,7 +176,9 @@ protected Stream<Optional<BuilderMapEntry>> createMilestones(
         create(config.getArrowGlacierBlockNumber(), specFactory.arrowGlacierDefinition(config)),
         create(config.getGrayGlacierBlockNumber(), specFactory.grayGlacierDefinition(config)),
         create(config.getMergeNetSplitBlockNumber(), specFactory.parisDefinition(config)),
-        create(config.getShandongBlockNumber(), specFactory.shandongDefinition(config)),
+        create(config.getShanghaiTime(), specFactory.shanghaiDefinition(config)),
+        create(config.getFutureTime(), specFactory.futureDefinition(config)),
+        create(config.getExperimentalTime(), specFactory.experimentalDefinition(config)),
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolScheduleBuilder.java:181]: Mainnet Time forks are created in TimestampScheduleBuilder, see https://github.com/hyperledger/besu/commit/3fd0681e3db862b2c90223a6bc9bfe81308f76d5 commit message for some explanation

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java [hunk 4]
```diff
@@ -94,7 +94,9 @@ default boolean isConsensusMigration() {
 
   OptionalLong getCancunTime();
 
-  OptionalLong getShandongBlockNumber();
+  OptionalLong getFutureTime();
```
Comment (siladu) [inline config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java:97]: Maybe we should be more explicit and use 'futureEipTime' to distinguish from the fact that 'experimentTime' can also be used as a non-EIP toggle?

**Review by siladu — COMMENTED** [review]
Mostly looks good, main issue is with timestamp fork creation which needs to be in TimestampScheduleBuilder.

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java (hunk 4 above)
Comment (jflo) [reply config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java:97]: yeah, FutureTime is a little too abstract. Maybe FutureForkTime? TBDForkTime?

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java [hunk 5]
```diff
@@ -96,11 +96,9 @@ public GenesisConfigOptions getConfigOptions(final Map<String, String> overrides
...
     final var optBaseFee = getBaseFeePerGas();
     if (optBaseFee.isPresent()) {
-      overridesRef =
-          Streams.concat(
-                  overrides.entrySet().stream(),
-                  Stream.of(Map.entry("baseFeePerGas", optBaseFee.get().toShortHexString())))
-              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+      // streams and maps cannot handle null values.
+      overridesRef = new HashMap<>(overrides);
+      overridesRef.put("baseFeePerGas", optBaseFee.get().toShortHexString());
```
Comment (jflo) [inline config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java:101]: way clearer

**Review by jflo — APPROVED** [review]

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java (hunk 4 above)
Comment (shemnon) [reply config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java:97]: This is why I proposed `Pizza` and `Bagal` ("The Works Pizza" for Future and "Everything Bagel" for Experimental)

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java [hunk 6]
```diff
@@ -94,7 +94,9 @@ default boolean isConsensusMigration() {
 
   OptionalLong getCancunTime();
 
-  OptionalLong getShandongBlockNumber();
+  OptionalLong getFutureEipsTime();
+
+  OptionalLong getExperimentalEipsTime();
```
Comment (siladu) [inline config/src/main/java/org/hyperledger/besu/config/GenesisConfigOptions.java:99]: ah ok, so experimental is intended to be for EIPs not as a general use feature flag? Makes sense I think, any non-EIP experimental feature probably warrants its own flag.

**Pull request merged**